### ISSUE TITLE
Publishing from the VS wizard walkthrough

### ISSUE THREAD

**Issue by SteveSandersonMS** [dotnet/aspire#3280]
This is based on **VS Version 17.10.0 Preview 2.0** and **Aspire 8.0.0-preview.4.24156.9**. My app is a `net8.0` Blazor Web app that makes use of a SQLite database, and a `RedisResource` which it starts up in Docker locally.

I don't know if the following is of any use to you all, nor whether the VS publishing flow is fully implemented and expected to work yet. Since it's in preview I fully appreciate that not all the details will be resolved yet, so maybe none of the following is news. Feel free to just close this if it doesn't contribute anything new (and no, I don't need you to debug my deployment for my own benefit - I'm only reporting this in case it's useful to the Aspire team).

### Walkthrough

Right-clicked on the `AppHost` project and chose *Publish*. The only option is ACA. I'm unsure if that's because it's the only option *ever* or if this is something specific to what components I'm using in my app:

![image](https://github.com/dotnet/aspire/assets/1101362/afe53eb0-afe6-410f-978d-d04e948a967d)

In any case, ACA is what I wanted so I'm happy with this. Clicked *Next*.

It suggests I might not have an Azure subscription, though I certainly do. In fact I "re-entered my credentials" less than 30 mins ago because this happened before.

![image](https://github.com/dotnet/aspire/assets/1101362/915bba14-f71a-49a9-93f3-9f788cc092d8)

I clicked the button to re-enter credentials, which flickered the UI for a second and then it shows the subscription (without asking for any credentials). And then another half-second later the warning to "re-enter credentials" reappears for some reason:

![image](https://github.com/dotnet/aspire/assets/1101362/254790b8-4ea4-443c-9ddb-861a4b17cde6)

I ignore that because it doesn't seem to stop me proceeding through the wizard.

I pick the subscription and deployment location, but have no idea what "Environment name" value to enter. **Request:** Could there be a help link here? Or one of those little "i" symbols that shows a tooltip with guidance?

![image](https://github.com/dotnet/aspire/assets/1101362/a2d5fe4e-3957-45c6-acf3-9b70cce1287f)

After some web searching, I enter the environment name `prod` and continue. It correctly detects the service I want to expose publicly. Great!

![image](https://github.com/dotnet/aspire/assets/1101362/08519e5c-66df-4603-a2f5-cd8718a09999)

I continue, thinking it might publish now, but it completes the wizard saying it created an environment.

![image](https://github.com/dotnet/aspire/assets/1101362/f06ce271-3f15-41a0-88bc-ac395decee85)

On the next dialog, I wait for this:

![image](https://github.com/dotnet/aspire/assets/1101362/759179dd-fd91-4eb4-9980-1ec32c83e1e6)

After a few minutes, it seems to complete successfully, so things are looking good. I click Publish, but after 10 seconds or so:

![image](https://github.com/dotnet/aspire/assets/1101362/aae2d087-13d2-425a-a391-92d5dba55d8a)

There's a link, "See logs in Output window", which highlights in white as you hover it and shows the "hand" pointer, clearly reaffirming that you can click it:

![image](https://github.com/dotnet/aspire/assets/1101362/32a91f1e-c293-4574-90d1-487948f7f57a)

However when you click it... nothing happens. Maybe it's because the Output window was already on the screen anyway? Maybe we could move focus to it or something. Or if something else was meant to happen I'm not sure what.

Checking the Output window manually:

```
Build started at 10:30...
1>------ Build started: Project: GitHubIssueSearch.ServiceDefaults, Configuration: Debug Any CPU ------
1>GitHubIssueSearch.ServiceDefaults -> C:\path\to\GitHubIssueSearch\GitHubIssueSearch.ServiceDefaults\bin\Debug\net8.0\GitHubIssueSearch.ServiceDefaults.dll
2>------ Build started: Project: GitHubIssueSearch.UI, Configuration: Debug Any CPU ------
2>GitHubIssueSearch.UI -> C:\path\to\GitHubIssueSearch\GitHubIssueSearch.UI\bin\Debug\net8.0\GitHubIssueSearch.UI.dll
3>------ Build started: Project: GitHubIssueSearch.AppHost, Configuration: Debug Any CPU ------
3>GitHubIssueSearch.AppHost -> C:\path\to\GitHubIssueSearch\GitHubIssueSearch.AppHost\bin\Debug\net8.0\GitHubIssueSearch.AppHost.dll
========== Build: 3 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Build completed at 10:30 and took 08.884 seconds ==========
29/03/2024 10:30:18: Info 
29/03/2024 10:30:18: Info Provisioning Azure resources (azd provision)
29/03/2024 10:30:18: Info Provisioning Azure resources can take some time.
29/03/2024 10:30:18: Info 
29/03/2024 10:30:18: Info Analyzing Aspire Application (this might take a moment...)
```

No indication of any failure, so at this point I'm unsure how to proceed.

### Summary

To reiterate, I'm aware this is preview and don't know which parts of this are meant to be fully-functional yet. Please don't feel you have to help me fix my deployment. I'm only reporting this in case any of the issues are surprising.

**Comment by SteveSandersonMS** [comment]
OK, so from poking around in Azure portal, I see clues indicating that I was using the wrong subscription. So I went through the above flow again using what I think is the correct subscription, and it's looking a lot better! It's pushing images to a container registry. And finally:

![image](https://github.com/dotnet/aspire/assets/1101362/97f1fb80-d5a1-4c79-a671-d03989193222)

Yay! So I click to go to the public `.azurecontainerapps.io` URL that it shows me in the publish output, hoping to see my web app. The browser spins its "loading" spinner for a long time. I don't know what's happening.

Eventually I give up waiting and go look in the Azure portal and find (and BTW the page is still "loading" in my other browser tab):

![image](https://github.com/dotnet/aspire/assets/1101362/c3d73a86-210f-4979-905d-33b1ff792e4d)

Several minutes later the page load fails:

![image](https://github.com/dotnet/aspire/assets/1101362/525baf41-719f-4e87-97d8-c3d9381992b4)

... but by this time I can see the problem is in my own application code. The ACA page in Azure portal gives pretty clear log output:

```
2024-03-29T11:16:55.398268974Z       An error occurred using the connection to database 'main' on server 'data/dotnet-runtime-issues.db'.
2024-03-29T11:16:55.410826247Z Unhandled exception. Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 14: 'unable to open database file'.
```

Yes, it's my fault - I didn't put the DB file in the right place so it would be deployed. I realise I'm not actually sure how one is meant to deploy SQLite - where is the `.db` file meant to live in prod, so that it doesn't get overwritten on each deployment? But this is a separate concern from Aspire.

**Comment by SteveSandersonMS** [comment]
After further investigation it seems unclear that there are any primitives available to ACA that would be suitable for a peristent SQLite database (e.g., see [discussion](https://github.com/MicrosoftDocs/azure-docs/issues/47130)).

**Comment by SteveSandersonMS** [comment]
I'm switching over to Postgres now and it works great locally, but my first deployment failed due to the `.db` file being readonly in prod. So I applied what I hope is a fix and click *Publish* in the Aspire UI again, and it republishes.

The updated deployment still fails at runtime, which is fine - I probably have still done something wrong - but what's hyper-frustrating is I can't see the console output to diagnose the problem. Here's the list of "revisions with issues":

![image](https://github.com/dotnet/aspire/assets/1101362/f02f8ffb-9a10-4796-81ee-a972ebd9ec4e)

Now, which of them is the old one and which is the new one? It doesn't say! So I click through the powjxqa one into its console logs and can see from those logs that it must be the old one (because I changed the log output in the new one).

So I go back and click through the d5m7yrs one and into its console logs and am super confused because it's also still the old console output. Did Aspire fail to deploy my change? No, look more closely:

![image](https://github.com/dotnet/aspire/assets/1101362/76a8c44b-5da6-4a9c-9643-6ffff9f1664c)

Even though I clicked the link from the d5m7yrs replica, the portal is showing the output from the powjxqa replica. OK, just a UI bug, right? So I'll change the replica from the dropdown:

![image](https://github.com/dotnet/aspire/assets/1101362/4ea79a4e-d2e9-4291-a063-488dea8c6e19)

Nope! For some reason I'm only allowed to see the powjxqa replica and hence can't diagnose what went wrong with d5m7yrs. Note that I re-checked that I clicked the right links multiple times, and found other paths through the portal UI to the logs, all of which also limited me to seeing the one from powjxqa.

Again, maybe my mental model is wrong here, but it's unclear how to diagnose this fault.

**Comment by bradygaster** [comment]
Thanks so much for the detailed write-up, @SteveSandersonMS. If you're on preview 2, try to get on the internal preview branch as we've resolved some of the earlier issues you had w.r.t. logging into azd on your command prompt. I've filed an issue to get the info bubble added about azd environments, too. Hadn't thought of that, but that's a great suggestion for folks new to the idea of azd environments. Do you think it'd be better if we defaulted the name? Something like `<projectname>2387dh` to randomize it?

**Comment by bradygaster** [comment]
@SteveSandersonMS (and everyone else) this error: 

`"permission to perform action 'Microsoft.Authorization/roleAssignments/write' at scope '/subscriptions/REDACTED_GUID/resourceGroups/rg-my-github-issue-search/providers/Microsoft.ContainerRegistry/registries/REDACTED_STRING/providers/Microsoft.Authorization/roleAssignments/REDACTED_GUID'.'"`

is typically due to the fact that your account lacks the appropriate Azure role required to assign `acrpull` rights to an Azure Container Registry. By the time you see this I will have fixed your access level on the test subscription I presume you're using. For customers using this feature, they'd need contributor rights to an Azure subscription, along with the roles appropriate for assigning the `acrpull` role to a container registry. This isn't an Aspire-specific thing, mind you, it's just that we're trying to enable a role that requires a role to be enabled. Note, this is only a requirement when you're creating the Azure resources, once they're there - say in the case of you having worked with a DevOps or platform engineering team to have your environment pre-created - you're just pushing updates to those container registries and to the associated Azure Container Apps. 

Also, w.r.t your container apps question, that's simply the first publishing target we've enabled in VS. We plan on enabling more (no time lines yet), but targeted Container Apps first as we feel ACA is a great hosting spot for Aspire apps.

**Comment by timheuer** [comment]
Great log of feedback @SteveSandersonMS! I think we've got some things already fixed in latest -- and some of these being Azure-specific we should see how we can surface them better if possible!  /cc @abpiskunov

**Comment by SteveSandersonMS** [comment]
OK, in the end it worked and my Aspire Blazor+SK+ONNX+Redis+Postgres app is finally running on ACA :)

Summary of learnings/questions:

 * Environment name could be clearer (@bradygaster filed as bug 2013595)
 * Permissions are painful :) We all knew that already.
 * **Possible bug:** The "See logs in Output window" link on the publish failure UI doesn't seem to do anything
   * Ideally it should cause the Output pane to be displayed and switched to the *Azure Developer CLI* tab, otherwise it's hard to discover that tab even exists. Do you think this is a bug, Brady?
 * Brady mentions the `azd login` issue should already be fixed on the next preview
 * AFAICT there's no way to use SQLite with persistence inside ACA, and hence inside Aspire - that's a bit of a tough limitation.
 * I'm unclear on good patterns for seeding a database in Aspire
   * We do have [docs about this](https://learn.microsoft.com/en-us/dotnet/aspire/database/seed-database-data?tabs=sql-server), but they only seem to address the question for local development. The docs' usage of volume mounts can't apply in prod because Aspire doesn't deploy volume mounts.
   * What I've done in my app is include a SQLite .db file with `PublishToOutputDirectory` so it gets deployed, then at runtime at app startup I copy it to a `/tmp` dir (because the app dir is readonly and hence can't open SQLite files there), and then loop over any rows not in the Postgres DB and copy them over. This is all pretty unpleasant, but would be less so if I was working with a smaller volume of data that I could just shove into a `.json` file.
   * More generally I'm unsure how DBs inside ACA are best managed in an ongoing way
 * So far the SK facilities are working fine in this environment, though questions about resilience are forcing me to change my mindset (i.e., all abstractions leak if they represent out-of-process calls) and it seems that there are still unmet gaps in some areas (e.g., making the Redis connector resilient).
 * **Bottom line:**
   * There are some really magic moments with Aspire. Realising I can add Postgres just by typing a few lines of C# and not having to touch any other tooling, write any yaml, or run any docker CLI commands, was awesome. Then wanting to add pgAdmin and realising that's only one more line of C# was amazing.
   * And yes I want the Aspire dashboard in prod like everyone else does - it's a vastly nicer way of seeing what's going on than anything else we have.

**Comment by timheuer** [comment]
>   * Ideally it should cause the Output pane to be displayed and switched to the _Azure Developer CLI_ tab, otherwise it's hard to discover that tab even exists. Do you think this is a bug, Brady?

Some improvements to raise visibility
![image](https://github.com/dotnet/aspire/assets/4821/c59f2e71-bcd2-47d2-8aae-76d8963af179)

> * I'm unclear on good patterns for seeding a database in Aspire

Been thinking about this too -- probably need some guidance on how EF migrations can help at least. 

>   * More generally I'm unsure how DBs inside ACA are best managed in an ongoing way

Is this a persistence concern? Obviously it depends on the db type, etc. some of the items you note on smaller dbs, or even pgsql probably using volume mounts to persist across container concerns would be smart...I'm still learning here myself.

> * **Bottom line:**
>   
>   * There are some really magic moments with Aspire. Realising I can add Postgres just by typing a few lines of C# and not having to touch any other tooling, write any yaml, or run any docker CLI commands, was awesome. Then wanting to add pgAdmin and realising that's only one more line of C# was amazing.

🎉🎉🎉

**Comment by bradygaster** [comment]
@SteveSandersonMS picking this one back up again - wanted to iterate on your points and address each. I'd like to know if you could re-test and validate my presumptions below. 

> Environment name could be clearer (@bradygaster filed as bug 2013595)

I'm not sure we changed anything here, but keen to see if you or others find this part confusing, still. 

> Permissions are painful :) We all knew that already.

Indeed. This is something I've asked the crew to think about - something like a pre-validation capability would be great, so we know if there exists any potential pending issues in the deployment the customer is about to try with suggestions on pre-mitigation. 

> Brady mentions the azd login issue should already be fixed on the next preview

Indeed, this was resolved. We're now passing the logged-in user in Visual Studio over to `azd`. 

> Possible bug: The "See logs in Output window" link on the publish failure UI doesn't seem to do anything. Ideally it should cause the Output pane to be displayed and switched to the Azure Developer CLI tab, otherwise it's hard to discover that tab even exists. Do you think this is a bug, Brady?

Can you try this again? It should open up the output window and show you the results and any errors `azd` runs into along the way. If indeed there exists a deployment issue that blocks deployment, I'd want us to show the output window on the error - or on success. 

> AFAICT there's no way to use SQLite with persistence inside ACA, and hence inside Aspire - that's a bit of a tough limitation. I'm unclear on good patterns for seeding a database in Aspire. We do have [docs about this](https://learn.microsoft.com/en-us/dotnet/aspire/database/seed-database-data?tabs=sql-server), but they only seem to address the question for local development. The docs' usage of volume mounts can't apply in prod because Aspire doesn't deploy volume mounts.

Is this still the case? Doesn't SQLite ship a containerized version one could pull down? cc @IEvangelist as the docs may have been updated in this space since the issue was created. 

> What I've done in my app is include a SQLite .db file with PublishToOutputDirectory so it gets deployed, then at runtime at app startup I copy it to a /tmp dir (because the app dir is readonly and hence can't open SQLite files there), and then loop over any rows not in the Postgres DB and copy them over. This is all pretty unpleasant, but would be less so if I was working with a smaller volume of data that I could just shove into a .json file.

I'll give this a try myself and report back; again, if this is this much of a pita I'd prefer we fix it, at a minimum ask @IEvangelist to find some time to document a workaround for other folks running into issues with SQLite.
